File: Unidad 4-Arreglos/Ejercicio7.js
/*
	Autor: Carlos Chan Gongora
	Entradas: El tamaño del vector.
	Salidas: La media, mediana y moda de los elementos del vector.
	Procedimiento general: Lee el largo del vector y posterior lo rellena con numeros aleatorios entre 0 y 50,
	despues calcula la media, moda, mediana y lo imprime.
*/

// Entradas
let vector = new Array();
let moda, tam;
let media = 0, mediana = 0;
let imprimir = "";
	
tam = parseInt(prompt("Ingrese el largo del vector: "));
	
// Proceso
rellenarVector(vector, tam);
media = calcularMedia(vector, tam);
moda = calcularModa(vector, tam);
mediana = calcularMediana(vector, tam);
// Salidas
imprimir = "Media: " + media + "\n" + "Mediana: " + mediana + "\n";
if(moda == -1){
	imprimir += "Moda: No hay moda\n";
}
else{
	imprimir += "Moda: " + moda;
}
alert(imprimir);
imprimirVector(vector, tam);


function rellenarVector(vec, tam){
	for(let i = 0; i < tam; i++){
		vec[i] = Math.round(Math.random() * 50);
	}
}

function calcularMedia(vec, tam){
	let media = 0;
	for(let i = 0; i < tam; i++){
		media += vec[i];
	}
	media = media / tam;
	return media;
}

function calcularMediana(vec, tam){
	let median = 0;
	let posicion = 0;
	// Si el largo del vector es par
	ordenarVector(vec, tam);
	if(tam % 2 == 0){
		posicion = (tam / 2) - 1;
		median = vec[posicion] + vec[posicion + 1];
		median = median / 2;
	}
	else{
	// si el largo del vector es impar
		posicion = ((tam + 1) / 2) - 1;
		median = vec[posicion];
	}
	
	return median;
}

function calcularModa(vec, tam){
 
	let moda = -1, veces = 0, ultimo_numero_veces = 0;//veces: contará el número de veces que se repite un número
	let j = 0;
	for(let i = 0; i < tam; i++){
		//j: analizará una posición adelante del valor actual
		j = i + 1;
		while(j < tam){
			//Si el número en i se repite más adelante, el número de veces incrementa
			if(vec[i] == vec[j]){
				veces++;
			}
			j++;
		}
 
		if(veces > ultimo_numero_veces){
			moda = vec[i];
			ultimo_numero_veces = veces;
			veces = 0;
		}
		else{
         veces = 0;
		}
	}
	//Devuelve el numero que más se repite
	return moda;
}

function ordenarVector(vec, tam){
	let aux = 0;
	for (let i = 0; i < tam - 1; i++){       
		for (let j = 0; j < tam - i - 1; j++){  
			if (vec[j] > vec[j+1]){
				aux = vec[j];
				vec[j] = vec[j + 1];
				vec[j + 1] = aux;
			} 
		}
   }
}

function imprimirVector(vec, tam){
	let impresion = "";
	for(let i = 0; i < tam; i++){
		impresion += vec[i] + ", ";
	}
	alert("Vector: " + impresion);
}

/* QA                                    
Por: Eyder Antonio Concha Moreno

Entradas: 4
Salidas: {16, 20, 34, 47, }
Ordenar vector: vector son ordenados apropiadamente - CORRECTO
Calcular media: 29.25 - CORRECTO
Calcular mediana: 27 - CORRECTO
Calcular Moda: No hay moda - CORRECTO

Entradas: 10
Salidas: {3, 10, 14, 16, 20, 26, 26, 33, 36, 40, }
Ordenar vector: vector son ordenados apropiadamente - CORRECTO
Calcular media: 22.4 - CORRECTO
Calcular mediana: 23 - CORRECTO
Calcular Moda: 26 - CORRECTO

Entradas: 7
Salidas: {0, 5, 16, 22, 23, 39, 48, }
Ordenar vector: vector son ordenados apropiadamente - CORRECTO
Calcular Media: 21.857142857142858 - CORRECTO
Calcular Mediana: 22 - CORRECTO
Calcular Moda: No hay moda - CORRECTO

Buen codigo 10/10
*/

File: Unidad 4-Arreglos/ejercicio7.py
import random

#Autor: Jose Eduardo Mendez Verdejo    09/Marzo/19
#Entradas: Un vector de tamano "n" con numeros random entre 0,50
#Salidas: la media, la moda y la mediana de ese vector
#Proceso: se define el tamano del vector, se le otorgan valores al vector mientras se suman para obetener la media
#		  la suma obtenida se divide entre el total de elementos "media / n", se pasa a obtener la moda, la cual
#		  se verifica cuantas veces se repite cada numero del 0-50, se analiza cualse repitio mas y en otro vector
#         se anaden los numeros que son la moda, para determinar la mediana hay que determinar si el tam de array 
#		  es par o impar, se determina la mediana  y si s par, se suman los puntos medios y se divide entre 2

#----- Genera los numeros random del vector -----
def generateArray(array,media):
	for i in range(0,n):
		array[i] = random.randrange(0,50)

		#se suman los numeros para calcular la media
		media = media + array[i]
	
	#se divine entre el numero de elementos del vector
	media = media / n

	return media	

#---- obtiene la moda -----
def getModa(trono):

	#determina la cantidad de veces que se repite cada numero de 0 a 50
	for j in range(0,50):
		for i in array:
			if(i == j):
				modaRepite[j] = modaRepite[j] + 1

	#determina cual fue el numero mayor de num. repetidos
	for i in modaRepite:
		if(i > trono):
			trono = i

	#determina los numeros que mas se repitieron y los añade al vector moda, cuales son la moda
	for i in range(0,50):
		if(modaRepite[i] == trono):
			moda.append(i) #añade en la ultima posicion el numero

#----- Obtiene la mediana -----
def getMediana(medianaPos1,medianaPos2):

	#determina si el tam del vector es par o impar
	if(n % 2 == 1):
		medianaPos1 = int(n/2) #se obtiene la pos de la mediana en caso impar
	else:

		#se obtiene la pos de las medianas en caso par
		medianaPos1 = int(n/2) - 1
		medianaPos2 = int(n/2) 

				
	if(medianaPos2 == 0):
		medianas = array[medianaPos1]
	else:
		mediana[0] = array[medianaPos1]
		mediana[1] = array[medianaPos2]	
		medianas = (mediana[0] + mediana[1]) / 2


	return medianas



n = int(input("Define le tamano del vector: "))

array = [None] * n
modaRepite = [0] * 50
moda = []
medianaPos1 = 0
medianaPos2 = 0
mediana = [0,0]
media = 0
trono = -1


media = generateArray(array,media)

getModa(trono)

medianas = getMediana(medianaPos1,medianaPos2)

# if(n % 2 == 1):
# 	medianaPos1 = int(n/2) 
# else:
# 	medianaPos1 = int(n/2) - 1
# 	medianaPos2 = int(n/2) 

			
# if(medianaPos2 == 0):
# 	mediana[0] = array[medianaPos1]
# else:
# 	mediana[0] = array[medianaPos1]
# 	mediana[1] = array[medianaPos2]	

print("Vector ",array)

print("la media es: ",media)

print("La moda es/son: ",moda)	

print("la mediana es:", medianas)

"""
Autor QA: Alejandro Torre Reyes
Entrada: 10
Salida: Vector  [10, 26, 43, 0, 16, 2, 41, 26, 9, 29]
la media es:  20.2
La moda es/son:  [26]
la mediana es: 9.0
Proceso: Ok. Todo correcto
"""
